fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/expressions/GroupingKeyExpression.java

```diff
@@ -167,6 +167,10 @@ public KeyExpression getGroupingSubKey() {
         return getWholeKey().getSubKey(0, getGroupingCount());
     }
 
+    public boolean isGroupingPermutable() {
+        return !createsDuplicates(); // TODO this can be optimized
```

Comment:
I'm having trouble formulating the theoretical reason that a single record contributing to more than one aggregate bucket would mean that permutations of the bucket keys aren't allowed. Is the real problem a lack of a solid way to specify conditions on such a key absent correlations? So this method really just means what it does, namely skip trying the particular permutation algorithm?

Reply:
That is absolutely correct. If I cannot take a `GroupingKeyExpression` apart into the grouping expressions (in order to permute them) in a lossless way (i.e. can I put them back together to get the original `GroupingKeyExpression`) I cannot do any permutations. As a consequence this method is used to skip that permuting algorithm (see `IndexFunctionHelper`)

Reply:
It might be that there is a better name that indicates the discretionary nature. Or maybe just making `INTERNAL` and adding a comment summarizing the reasoning is enough.